src/Microsoft.Azure.NotificationHubs/KeyValueConfigurationManager.cs

```diff
@@ -106,10 +122,208 @@ private static NameValueCollection CreateNameValueCollectionFromConnectionString
 
         public void Validate()
         {
-            if (string.IsNullOrWhiteSpace(this.connectionProperties[EndpointConfigName]))
+            if (string.IsNullOrWhiteSpace(connectionProperties[EndpointConfigName]))
             {
                 throw new ConfigurationException(SRClient.AppSettingsConfigMissingSetting(EndpointConfigName, ServiceBusConnectionKeyName));
             }
         }
+
+        public NamespaceManager CreateNamespaceManager()
+        {
+            Validate();
+
+            string operationTimeout = connectionProperties[OperationTimeoutConfigName];
+            IEnumerable<Uri> endpoints = GetEndpointAddresses(connectionProperties[EndpointConfigName], connectionProperties[ManagementPortConfigName]);
+            IEnumerable<Uri> stsEndpoints = GetEndpointAddresses(connectionProperties[StsEndpointConfigName], null);
+            string issuerName = connectionProperties[SharedSecretIssuerConfigName];
+            string issuerKey = connectionProperties[SharedSecretValueConfigName];
+            string sasKeyName = connectionProperties[SharedAccessKeyName];
+            string sasKey = connectionProperties[SharedAccessValueName];
+            string windowsDomain = connectionProperties[WindowsDomainConfigName];
+            string windowsUsername = connectionProperties[WindowsUsernameConfigName];
+            SecureString windowsPassword = GetWindowsPassword();
+            string oauthDomain = connectionProperties[OAuthDomainConfigName];
+            string oauthUsername = connectionProperties[OAuthUsernameConfigName];
+            SecureString oauthPassword = GetOAuthPassword();
+
+            try
+            {
+                TokenProvider provider = CreateTokenProvider(stsEndpoints, issuerName, issuerKey, sasKeyName, sasKey, windowsDomain, windowsUsername, windowsPassword, oauthDomain, oauthUsername, oauthPassword);
+                if (string.IsNullOrEmpty(operationTimeout))
+                {
+                    return new NamespaceManager(endpoints, provider);
+                }
+
+                return new NamespaceManager(
+                    endpoints,
+                    new NamespaceManagerSettings()
+                    {
+                        TokenProvider = provider
+                    });
+            }
+            catch (ArgumentException e)
+            {
+                throw new ConfigurationErrorsException(
+                    SRClient.AppSettingsCreateManagerWithInvalidConnectionString(e.Message),
+                    e);
+            }
+            catch (UriFormatException e)
+            {
+                throw new ConfigurationErrorsException(
+                    SRClient.AppSettingsCreateManagerWithInvalidConnectionString(e.Message),
+                    e);
+            }
+        }
+
+        public SecureString GetWindowsPassword()
+        {
+            return GetSecurePassword(WindowsPasswordConfigName);
+        }
+
+        public SecureString GetOAuthPassword()
```

Comment:
Are APIs like this required to be public or can they be private?

Reply:
changed to private